src/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/SocketConnection.cs

```diff
@@ -0,0 +1,213 @@
+        {
+            Debug.Assert(socket != null);
+            Debug.Assert(transport != null);
+
+            _socket = socket;
+            _transport = transport;
+
+            _localEndPoint = (IPEndPoint)_socket.LocalEndPoint;
+            _remoteEndPoint = (IPEndPoint)_socket.RemoteEndPoint;
+        }
+
+        public async void Start(IConnectionHandler connectionHandler)
+        {
+            IConnectionContext context = connectionHandler.OnConnection(this);
+
+            _input = context.Input;
+            _output = context.Output;
+
+            // Spawn send and receive logic
+            Task receiveTask = DoReceive();
+            Task sendTask = DoSend();
+
+            // Wait for them to complete (note they won't throw exceptions)
+            await receiveTask;
+            await sendTask;
+
+            _socket.Dispose();
+        }
+        
+        private async Task DoReceive()
+        {
+            try
+            {
+                bool done = false;
+                while (!done)
+                {
+                    // Ensure we have some reasonable amount of buffer space
+                    WritableBuffer buffer = _input.Alloc(MinAllocBufferSize);
+
+                    try
+                    {
+                        int bytesReceived = await _socket.ReceiveAsync(GetArraySegment(buffer.Buffer), SocketFlags.None);
+
+                        if (bytesReceived == 0)
+                        {
+                            // EOF
+                            _input.Complete();
+                            done = true;
+                            break;
+                        }
+
+                        // record what data we filled into the buffer and push to pipe
+                        buffer.Advance(bytesReceived);
+                    }
+                    finally
+                    {
+                        var result = await buffer.FlushAsync();
+                        if (result.IsCompleted)
+                        {
+                            // Pipe consumer is shut down
+                            _socket.Shutdown(SocketShutdown.Receive);
+                            done = true;
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                _input.Complete(ex);
+            }
+        }
+
+        private void SetupSendBuffers(ReadableBuffer buffer)
+        {
+            Debug.Assert(!buffer.IsEmpty);
+            Debug.Assert(!buffer.IsSingleSpan);
+
+            if (_sendBufferList == null)
+            {
+                _sendBufferList = new List<ArraySegment<byte>>();
+            }
+
+            // We should always clear the list after the send
+            Debug.Assert(_sendBufferList.Count == 0);
+
+            foreach (var b in buffer)
+            {
+                _sendBufferList.Add(GetArraySegment(b));
+            }
+        }
+
+        private async Task DoSend()
```

Comment:
```
Failed   EmitsConnectionStartAndStop
Error Message:
 The collection contained 0 matching element(s) instead of 1.
Stack Trace:
   at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.EventSourceTests.<EmitsConnectionStartAndStop>d__2.MoveNext()
```

I haven't debugged, will after I've made the other changes here.